docs/developer/logging-guidelines.md

```diff
@@ -0,0 +1,407 @@
+4. Use camelCase for key names.
+5. When dealing with a Kubernetes resource, use `resource` as a key:
+
+   ```go
+    logger.Info("processed resource",
+        "resource", hr)
+   ```
+
+6. Don't put structs as values expect for Kubernetes resources, to prevent messy and potentially unbounded output.
+
+### Log Formatting
+
+By default, log messages are formatted as JSON, so that external systems can easily parse them.
+
+For the developer, log messages are formatted as text strings (except key/values), for human-friendly parsing. For
+example:
+
+```text
+2023-07-21T12:41:37.640-0400    INFO    processed resource      {"resource": {"apiVersion": "gateway.networking.k8s.io/v1beta1", "kind": "HTTPRoute", "namespace": "test", "name": "hr-1"}}
+```
+
+The formatting is controlled during the logger initialization.
+
+### When to Log
+
+Below are the cases for when to log a message:
+
+- Lifecycle events
+  - The control plane starts or shuts down
+  - Lifecycle of the components.
+- Handling a transaction/request
+  - The main transaction in the control plane is processing a batch of events related to Kubernetes resources taken
+      from the main event loop.
+  - Additional transactions could be out of the event loop requests to auxiliary control plane endpoints (APIs).
+- Interaction with external components. When the control plane interacts with an external component, log the stages of
+  that interaction as well as the result of it.
+
+For those cases:
+
+- Log an error only if the code doesn't handle it (performs some action). If it is handled, but it is still important to
+  record the fact of that
+  error, log it as an info message.
+- Consider including performance-related information. For example, the duration of an operation.
+- Log extra details (or steps) to facilitate troubleshooting under the `debug` level.
+
+> It is important to not overload the `info` level with too many details and messages.
+
+### What not to Log
+
+- *Sensitive information*, including the contents of the secrets, passwords, etc.
+- *Returned errors*. If a function returns an error, do not log it inside that function, because it will be logged or
+  handled
+  up in the stack.
+- *Critical errors*, for which we [panic](go-style-guide.md#use-panics-for-unrecoverable-errors-or-programming-errors).
+
+### Performance
+
+Logging too frequently can lead to poor performance, because of:
+
+- Too many I/O operations.
+- Too many allocations related to preparing the log message.
+
+If we diligently follow the guidelines from this document, the `error` and `info` log levels will not have many messages
+(many thousands per second). At the same time, such a case might still be possible, for example, in the case of a bug.
+To prevent that, the default settings in the controller-runtime helpers for zap
+[configure sampling][sampling].
+The sampling is based on the message (the `msg` parameter) of the log-related methods. Read more about sampling
```

Comment:
Because those helpers make such choices that affect us directly, and they might change it behind the scene, I wonder if we should use them.